### PR TITLE
some preliminary containers that can run cryptography's tests

### PR DESCRIPTION
**Pull request by reaperhulk**
No doubt we can improve these dramatically, but here's some early work (these docker containers work for compiling and testing cryptography)

### CODE REVIEW

File: build_images.sh [hunk 1]
```diff
@@ -7,3 +7,22 @@ popd
 pushd caddy
 docker build -t caddy .
 popd
+
+pushd builders
+pushd centos7
+docker build -t cryptography-centos7 .
```
Comment (alex) [inline build_images.sh]: I think the names should use `cryptography-runner` as a prefix or something else to clearly indicate they're for running jobs.

File: builders/centos7/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ARG user=jenkins
+ARG group=jenkins
+ARG uid=1000
+ARG gid=1000
+
+# Docker pipeline plugin mounts the workspace into the image
+# This adds the same uid/gid that we use in the jenkins container
+# so getpwuid() won't fail when tox invokes
+RUN groupadd -g ${gid} ${group} \
+    && useradd -u ${uid} -g ${gid} -m -s /bin/bash ${user}
```
Comment (alex) [inline builders/centos7/Dockerfile]: These can be two separate `RUN` commands.

File: builders/jessie/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ARG user=jenkins
+ARG group=jenkins
+ARG uid=1000
+ARG gid=1000
+
+# Docker pipeline plugin mounts the workspace into the image
+# This adds the same uid/gid that we use in the jenkins container
+# so getpwuid() won't fail when tox invokes
+RUN groupadd -g ${gid} ${group} \
+    && useradd -u ${uid} -g ${gid} -m -s /bin/bash ${user}
```
Comment (alex) [inline builders/jessie/Dockerfile]: Same

File: builders/libressl/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,38 @@
+FROM debian:jessie
```
Comment (alex) [inline builders/libressl/Dockerfile]: Name this `libressl-portable` or `libressl-jessie` or something?

File: builders/libressl/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ARG user=jenkins
+ARG group=jenkins
+ARG uid=1000
+ARG gid=1000
+
+# Docker pipeline plugin mounts the workspace into the image
+# This adds the same uid/gid that we use in the jenkins container
+# so getpwuid() won't fail when tox invokes
+RUN groupadd -g ${gid} ${group} \
+    && useradd -u ${uid} -g ${gid} -m -s /bin/bash ${user}
```
Comment (alex) [inline builders/libressl/Dockerfile]: Same

File: builders/libressl/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+RUN apt-get -qq update
+
+
```
Comment (alex) [inline builders/libressl/Dockerfile]: Nit: one blank line.

File: builders/libressl/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ARG LIBRE_VERSION=2.5.3
+
+RUN curl -s -O https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-${LIBRE_VERSION}.tar.gz
+
+RUN tar zxvf libressl-${LIBRE_VERSION}.tar.gz
+
+WORKDIR libressl-${LIBRE_VERSION}
+
+RUN ./config -Wl -Wl,-Bsymbolic-functions -fPIC shared --prefix=/usr/local/libressl && make -j4 install
```
Comment (alex) [inline builders/libressl/Dockerfile]: Clean up the source dir after we compile?

File: builders/stretch/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ARG user=jenkins
+ARG group=jenkins
+ARG uid=1000
+ARG gid=1000
+
+# Docker pipeline plugin mounts the workspace into the image
+# This adds the same uid/gid that we use in the jenkins container
+# so getpwuid() won't fail when tox invokes
+RUN groupadd -g ${gid} ${group} \
+    && useradd -u ${uid} -g ${gid} -m -s /bin/bash ${user}
```
Comment (alex) [inline builders/stretch/Dockerfile]: Same

File: builders/libressl/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    libffi-dev \
+    python-dev \
+    git \
+    curl
+
+ARG LIBRE_VERSION=2.5.3
+
+RUN curl -s -O https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-${LIBRE_VERSION}.tar.gz
+
+RUN tar zxvf libressl-${LIBRE_VERSION}.tar.gz
```
Comment (alex) [inline builders/libressl/Dockerfile]: Drop the `v`, makes the stdout of building the container unrearably large

File: builders/libressl/Dockerfile (hunk 4 above)
Comment (reaperhulk) [inline builders/libressl/Dockerfile]: (Now tagged cryptography-runner:jessie-libressl-$version). I can rename the dir as well if you want.

**Comment by alex** [comment]
builders directory should probably be renamed `runners`?

File: builders/libressl/Dockerfile (hunk 4 above)
Comment (alex) [inline builders/libressl/Dockerfile]: Renaming the dir would be aogod

File: builders/libressl/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+
+RUN apt-get install -qq -y \
+    build-essential \
+    libffi-dev \
+    python-dev \
+    git \
+    curl
+
+ARG LIBRE_VERSION=2.5.3
```
Comment (alex) [inline builders/libressl/Dockerfile]: Is there a reason to have a default here, given we'll parameterize it twice?

File: builders/libressl/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+RUN curl -s -O https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-${LIBRE_VERSION}.tar.gz
+
+RUN tar zxf libressl-${LIBRE_VERSION}.tar.gz
+
+WORKDIR libressl-${LIBRE_VERSION}
+
+RUN ./config -Wl -Wl,-Bsymbolic-functions -fPIC shared --prefix=/usr/local/libressl && make -j4 install
```
Comment (alex) [inline builders/libressl/Dockerfile]: Split into two `RUN`

File: builders/libressl/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+WORKDIR ../
+
+RUN rm -rf libressl*
```
Comment (alex) [inline builders/libressl/Dockerfile]: Surely this can be done without a glob?

File: build-images.py [hunk 13]
```diff
@@ -1,22 +1,60 @@
...
 import subprocess
 
 
+Image = collections.namedtuple("Image", ["tag", "path", "build_args"])
+
 IMAGES = [
-    ("jenkins", "crypto-jenkins"),
-    ("caddy", "caddy"),
+    Image(
+        tag="crypto-jenkins", path=["jenkins"], build_args=None
```
Comment (alex) [inline build-images.py]: Should `build_args` be an empty list?

File: runners/jessie-libressl/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+RUN tar zxf libressl-${LIBRE_VERSION}.tar.gz
+
+WORKDIR libressl-${LIBRE_VERSION}
+
+RUN ./config -Wl -Wl,-Bsymbolic-functions -fPIC shared --prefix=/usr/local/libressl && make -j4 install
+
+WORKDIR ../
+
+RUN rm -rf libressl-${LIBRE_VERSION} && rm libressl-${LIBRE_VERSION}.tar.gz
```
Comment (alex) [inline runners/jessie-libressl/Dockerfile]: You don't need two `rm` invocations, one will do.

**Pull request merged**